IPython/terminal/ptshell.py

```diff
@@ -317,5 +324,67 @@ def enable_gui(self, gui=None):
         else:
             self._inputhook = None
 
+    # Methods to support auto-editing of SyntaxErrors:
+
+    def edit_syntax_error(self):
+        """The bottom half of the syntax error handler called in the main loop.
+
+        Loop until syntax error is fixed or user cancels.
+        """
+
+        while self.SyntaxTB.last_syntax_error:
+            # copy and clear last_syntax_error
+            err = self.SyntaxTB.clear_err_state()
+            if not self._should_recompile(err):
+                return
+            try:
+                # may set last_syntax_error again if a SyntaxError is raised
+                self.safe_execfile(err.filename, self.user_ns)
+            except:
+                self.showtraceback()
+            else:
+                try:
+                    f = open(err.filename)
```

Comment:
Why not. This is copy-pasted code from the old TIS class, so I don't want to change it too much, but putting this in a `with` is a simple improvement.